src/main/java/net/minecraftforge/registries/ForgeRegistryTagManager.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Minecraft Forge - Forge Development LLC
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.registries;
+
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Sets;
+import net.minecraft.core.HolderSet;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.TagKey;
+import net.minecraftforge.registries.tags.IReverseTag;
+import net.minecraftforge.registries.tags.ITag;
+import net.minecraftforge.registries.tags.ITagManager;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+class ForgeRegistryTagManager<V extends IForgeRegistryEntry<V>> implements ITagManager<V>
+{
+    private final ForgeRegistry<V> owner;
+    private volatile Map<TagKey<V>, ITag<V>> tags = new IdentityHashMap<>();
+
+    ForgeRegistryTagManager(ForgeRegistry<V> owner)
+    {
+        this.owner = owner;
+    }
+
+    void bind(Map<TagKey<V>, HolderSet.Named<V>> holderTags)
+    {
+        IdentityHashMap<TagKey<V>, ITag<V>> newTags = new IdentityHashMap<>(this.tags);
+        Set<TagKey<V>> bound = new HashSet<>();
```

Comment:
Is there a reason each element is added to a set tracking the bound tags instead of just using `holderTags#keySet` for comparing against the previously bound tags?